components/script/script_thread.rs

```diff
@@ -2731,23 +2732,29 @@ impl ScriptThread {
         match window {
             None => warn!("postMessage after target pipeline {} closed.", pipeline_id),
             Some(window) => {
-                // FIXME: synchronously talks to constellation.
-                // send the required info as part of postmessage instead.
-                let source = match self.window_proxies.remote_window_proxy(
-                    &self.senders,
-                    window.upcast::<GlobalScope>(),
-                    source_browsing_context,
-                    source_pipeline_id,
-                    None,
-                ) {
-                    None => {
-                        return warn!(
-                            "postMessage after source pipeline {} closed.",
-                            source_pipeline_id,
-                        );
-                    },
-                    Some(source) => source,
-                };
+                let mut last: Option<DomRoot<WindowProxy>> = None;
+                for browsing_context_id in source_with_ancestry.into_iter().rev() {
+                    if let Some(window_proxy) = self.window_proxies.get(browsing_context_id) {
+                        last = Some(window_proxy);
+                        continue;
+                    }
+                    let window_proxy = WindowProxy::new_dissimilar_origin(
+                        window.upcast::<GlobalScope>(),
+                        browsing_context_id,
+                        source_webview,
+                        last.as_deref(),
+                        None,
+                        CreatorBrowsingContextInfo::from(last.as_deref(), None),
+                    );
+                    self.window_proxies
+                        .insert(browsing_context_id, window_proxy.clone());
+                    last = Some(window_proxy);
+                }
+
+                // Step 8.3: Let source be the WindowProxy object corresponding to
+                // incumbentSettings's global object (a Window object).
+                let source = last.expect("Source with ancestry should contain at least one bc.");
```

Comment:
should this be early return?

Reply:
That there is at least one bc in the vec is an invariant; so no because it would mean something is wrong beyond repair.